tests/exchange/test_exchange.py

```diff
@@ -3288,3 +3288,316 @@ def test_get_max_leverage(default_conf, mocker, pair, nominal_value, max_lev):
+        time: 10, mark_price: 2.84, nominal_value: 85.19999999999999, fundingRate: 0.000013, fundingFee: 0.0011075999999999998
+        time: 11, mark_price: 2.81, nominal_value: 84.3, fundingRate: 0.000077, fundingFee: 0.0064911
+        time: 12, mark_price: 2.81, nominal_value: 84.3, fundingRate: 0.000072, fundingFee: 0.0060696
+        time: 13, mark_price: 2.82, nominal_value: 84.6, fundingRate: 0.000097, fundingFee: 0.008206199999999999
+
+    contract_size: 50
+        time: 0, mark_price: 2.77, nominal_value: 138.5, fundingRate: -0.000008, fundingFee: -0.001108
+        time: 1, mark_price: 2.73, nominal_value: 136.5, fundingRate: -0.000004, fundingFee: -0.0005459999999999999
+        time: 2, mark_price: 2.74, nominal_value: 137.0, fundingRate: 0.000012, fundingFee: 0.001644
+        time: 3, mark_price: 2.76, nominal_value: 138.0, fundingRate: -0.000003, fundingFee: -0.00041400000000000003
+        time: 4, mark_price: 2.76, nominal_value: 138.0, fundingRate: -0.000007, fundingFee: -0.000966
+        time: 5, mark_price: 2.77, nominal_value: 138.5, fundingRate: 0.000003, fundingFee: 0.0004155
+        time: 6, mark_price: 2.78, nominal_value: 139.0, fundingRate: 0.000019, fundingFee: 0.002641
+        time: 7, mark_price: 2.78, nominal_value: 139.0, fundingRate: 0.000003, fundingFee: 0.000417
+        time: 8, mark_price: 2.77, nominal_value: 138.5, fundingRate: -0.000003, fundingFee: -0.0004155
+        time: 9, mark_price: 2.77, nominal_value: 138.5, fundingRate: 0, fundingFee: 0.0
+        time: 10, mark_price: 2.84, nominal_value: 142.0, fundingRate: 0.000013, fundingFee: 0.001846
+        time: 11, mark_price: 2.81, nominal_value: 140.5, fundingRate: 0.000077, fundingFee: 0.0108185
+        time: 12, mark_price: 2.81, nominal_value: 140.5, fundingRate: 0.000072, fundingFee: 0.010116
+        time: 13, mark_price: 2.82, nominal_value: 141.0, fundingRate: 0.000097, fundingFee: 0.013677
+    '''
+    d1 = datetime.strptime(f"{d1} +0000", '%Y-%m-%d %H:%M:%S %z')
+    d2 = datetime.strptime(f"{d2} +0000", '%Y-%m-%d %H:%M:%S %z')
+    api_mock = MagicMock()
+    api_mock.fetch_funding_rate_history = MagicMock(return_value=funding_rate_history)
+    api_mock.fetch_ohlcv = MagicMock(return_value=mark_ohlcv)
+    type(api_mock).has = PropertyMock(return_value={'fetchOHLCV': True})
+    type(api_mock).has = PropertyMock(return_value={'fetchFundingRateHistory': True})
+
+    exchange = get_patched_exchange(mocker, default_conf, api_mock, id=exchange)
+    funding_fees = exchange.calculate_funding_fees('ADA/USDT', amount, d1, d2)
+    assert funding_fees == expected_fees
+
+
+def test_calculate_funding_fees_datetime_called(
+    mocker,
+    default_conf,
+    funding_rate_history,
+    mark_ohlcv
+):
+    api_mock = MagicMock()
+    api_mock.fetch_ohlcv = MagicMock(return_value=mark_ohlcv)
+    api_mock.fetch_funding_rate_history = MagicMock(return_value=funding_rate_history)
+    datetime = MagicMock()
+    datetime.now = MagicMock()
```

Comment:
I'm trying to test that this branch get's executed

```
if not close_date:
            close_date = datetime.now(timezone.utc)
```

But I don't know what the best way to add this magic mock to `get_patched_exchange` would be

Reply:
Also, I'm getting this error

```
    def test_calculate_funding_fees_datetime_called(
        mocker,
        default_conf,
        funding_rate_history,
        mark_ohlcv
    ):
        api_mock = MagicMock()
        api_mock.fetch_ohlcv = MagicMock(return_value=mark_ohlcv)
        api_mock.fetch_funding_rate_history = MagicMock(return_value=funding_rate_history)
        datetime = MagicMock()
        datetime.now = MagicMock()
        type(api_mock).has = PropertyMock(return_value={'fetchOHLCV': True})
        type(api_mock).has = PropertyMock(return_value={'fetchFundingRateHistory': True})
    
        # TODO-lev: Add datetime MagicMock
        exchange = get_patched_exchange(mocker, default_conf, api_mock)
>       exchange.calculate_funding_fees('ADA/USDT', 30.0, datetime("2021-09-01 00:00:00"))
E       TypeError: calculate_funding_fees() missing 1 required positional argument: 'close_date'
```

which doesn't make any sense because close_date is optional

```
 def calculate_funding_fees(
        self,
        pair: str,
        amount: float,
        open_date: datetime,
        close_date: Optional[datetime]
    ) -> float:
```

Reply:
you did define `close_date` as Optional - but without default attribute - which means you'll have to explicitly pass `None` as that parameter. What you probably wanted to do is `close_date: Optional[datetime] = None` ... ?

Reply:
As for the test - don't "magicMock" this, but use time-machine (if you rely on the date you're receiving).

Alternatively, ensure that "current time" will work for the test by moving other dates into the past/future (we do this in other tests, specifically with pairLocks ...)